### ISSUE TITLE
XLabs Sampler very slow

### ISSUE THREAD

**Issue by lchauvin** [XLabs-AI/x-flux-comfyui#126]
Hello,

I'm trying to run the IP Adapter workflow, but the XLabs Sampler is extremely slow. It takes more than 1h to generate 1 image.

I'm running on a 4070 Ti 12GB and I'm using the fp8 model.

Is that normal ?

**Comment by lchauvin** [comment]
I tested with a SampleCustomAdvanced with a FluxGuidance node, and it's much faster (about 60s to generate an image). Not sure why the XLabs Sampler is so slow then.

**Comment by em2er** [comment]
Same problem, 4060 with 8GB (GGUF Q4 dev), estimated time about 1 hour for 20 steps. Also I tried with default KSampler with XLab Ipadapter, it takes about 1-2 minutes but gives some vague results.

**Comment by lchauvin** [comment]
In the end, I switched to KSampler, removed IPAdapter, and used controlnets (but I have to make sure controlnet steps don't overlap (like Controlnet 1: 0.0 - 0.3, Controlnet 2: 0.3 - 0.6 for example. If I try to run them both at the same time like Controlnet1: 0.0 - 0.3, Controlnet 2: 0.0 - 0.3 it gets really slow and results are bad). I think the slowness might be due to the fact that it tries to load all controlnet at the same time, but I don't have enough VRAM.

Using GGUF Q8, I can use controlnets and loras and just reach the maximum VRAM (sometimes I have to close everything else, otherwise it's really slow), but it works. I also heard to use `--reserve-vram 1.2` at startup to help. I'm doing it and it works, but I'm not sure it's related.

**Comment by enesaltinkaya** [comment]
When I set **timestep_to_start_cfg**  equals or more than the **steps**, it runs as fast as ksampler.  
  
You know how we set cfg to 1 when using flux with ksampler, timestep_to_start_cfg must be setting it more than 1, so it slows down.

**Comment by lchauvin** [comment]
> When I set **timestep_to_start_cfg** equals or more than the **steps**, it runs as fast as ksampler.
> 
> You know how we set cfg to 1 when using flux with ksampler, timestep_to_start_cfg must be setting it more than 1, so it slows down.

That's interesting !!

**Comment by em2er** [comment]
> When I set **timestep_to_start_cfg** equals or more than the **steps**, it runs as fast as ksampler.
> 
> You know how we set cfg to 1 when using flux with ksampler, timestep_to_start_cfg must be setting it more than 1, so it slows down.

It's peculiar, but when I connect Canny ControlNet to the XLabs Sampler, the IPAdapter on the GGUF Q4 development model takes approximately 3 minutes for 15 steps, around 6 min for 30 steps, and about 8 min for 40 steps, even with **timestep_to_start_cfg set to 1** (or any other value). The ControlNet strength is 0.0. However, when I bypass the ControlNet, the process takes hours to complete just 15 steps.

**Comment by enesaltinkaya** [comment]
> > When I set **timestep_to_start_cfg** equals or more than the **steps**, it runs as fast as ksampler.
> > You know how we set cfg to 1 when using flux with ksampler, timestep_to_start_cfg must be setting it more than 1, so it slows down.
> 
> It's peculiar, but when I connect Canny ControlNet to the XLabs Sampler, the IPAdapter on the GGUF Q4 development model takes approximately 3 minutes for 15 steps, around 6 min for 30 steps, and about 8 min for 40 steps, even with **timestep_to_start_cfg set to 1** (or any other value). The ControlNet strength is 0.0. However, when I bypass the ControlNet, the process takes hours to complete just 15 steps.

I won't be able to present any opinions i'm afraid.  
I cant remember why i installed xlabs in the first place.  
But when i noticed the sampling speed difference with timestep_to_start_cfg i wanted to share :p

**Comment by ronaldomirandah** [comment]
its by default set to CPU, and it will take forever.
Change to GPU and it will run in less than 1 minute (RTX 3060 12gb) here!

**Comment by lchauvin** [comment]
> its by default set to CPU, and it will take forever. Change to GPU and it will run in less than 1 minute (RTX 3060 12gb) here!

How do you change it to gpu ?

**Comment by Signynt** [comment]
@lchauvin did you figure out how to change it to the gpu?
@ronaldomirandah Could you maybe explain how?

**Comment by lchauvin** [comment]
No, but I find hard to believe that it would run on cpu by default
